### PR TITLE
support fromstring bytea

### PR DESCRIPTION
**Pull request by bjne**
Support binary data by adding bytea support to fromstring

### CODE REVIEW

**Comment by eugwne** [comment]
I see here that pull request changed casting for bytea from the standart postgres cast to a custom cast bytea and so fromstring('bytea', "\\001\\001\\001") will not be the same as select CAST(E'\\001\\001\\001' AS bytea).
Enum changed behaviour where request is about 'support fromstring bytea' (do we need it? discussion? why here and not in the other request or issue?).

**Comment by bjne** [comment]
I made the common mistake of creating a pull request from master, and added the enum tostring
function later. Was not ment to be part of this pull request. I can create another one, or you can
just apply the bytea part?

wrt enum string, I know it is breaking compability... (but who wants oid enums anyway?)

**Comment by eugwne** [comment]
lua fromstring(type, string) == pg cast(string as type) . This pull request breaks this for type 'bytea'. conversion from bytes to bytea might be implemented without breaking changes here or in a custom library. I don't see any reason why to apply any breaking changes

**Comment by bjne** [comment]
Goal here was to support binary data from lua, and current bytea comversion does not support
that as far as I can see. I am not sure what you mean with fromstring(type, string) == pg_cast(string as type) .. I think there is a difference here, because a bytea can have \0's in it?, and is not really a string in C's concept, but a lua string is fine with it.

**Comment by eugwne** [comment]
By fromstring(type, string) == pg_cast(string as type)  I meant that fromstring function is equal to a postgres cast operator, they make the same and results should be the same. In the issue #58 I see an attempt to use fromstring('bytea', "\x01\x00\x01") and what I see if I want to get 0x01 0x00 0x01 values I use something like
```sql
select encode(CAST(E'\\x010001' AS bytea), 'hex')
``` 
and lua output is the same for fromstring
```lua
do $$  
local encode = pgfunc("encode(bytea, text)")
local s = fromstring('bytea', "\\x010001")
print(encode(s, 'hex'))
$$ language pllua
```

or I different way
```sql
select encode(CAST(E'\\001\\000\\001' AS bytea), 'hex')
```
and lua shows the same
```lua
do $$  
local encode = pgfunc("encode(bytea, text)")
local s = fromstring('bytea', "\\001\\000\\001")
print(encode(s, 'hex'))
$$ language pllua
```
but in the issue fromstring('bytea', "\x01\x01\x01"), than it should work as 
```sql
select encode(CAST(E'\x01\x00\x01' AS bytea), 'hex')
```
but than doesn't work and the output is ERROR:  invalid byte sequence for encoding "UTF8": 0x00
so fromstring('bytea', "\x01\x01\x01") shouldn't work at all, but it gets from lua only bytes before 0x00 and postgres doesn't get 0x00 and that's why does not complain adout  invalid byte sequence for encoding.

So if we use this changes which just copy bytes, than  '\x01\x00\x01' converted to [01 00 01] but "\\001\\000\\001'' make something like [5c 30 30 31 5c 30 30 30 5c 30 30 31] (I haven't checked, but looks like that, and it is not what is expected)
To summarize: fromstring suppors conversion for byte array (almost as expected but doesn't throw an error that string(text or wharever readable) expected, not bytes). Pull request makes fromstring work with raw bytes (and only with them)for bytea instead of text (shall we make fromraw or frombytes? or special type for bytearray?)

**Comment by bjne** [comment]
I am not sure. I think this really boils down to what "string" in fromstring _is_, if it is a luastring, I think
my approach is correct, the luastring _can_ contain \0 and fromstring('bytea', luastring) does give the
expected result. If you mean that string is a C str without len, we will have to implement different call.

I was trying to just do this transparent if a function returned a luastring when the return type was bytea,
but did not manage to make it work..

**Comment by eugwne** [comment]
fromstring is conversion function that uses postgres registered functions and only them to convert values, not a memcpy.

**Comment by bjne** [comment]
So what do you want to do to support conversion from lua-string (actually a bytea) to pg bytea?

**Comment by eugwne** [comment]
fromstring uses postgres input function to convert values. If  postgres input function gets wrong input  (whatever it is, pointer, cstring, lua-string ...) than that should be fixed, if not then a custom type conversion should be registered in postgres. Changes for **fromstring**  with not using  postgres input function will not be applied.

**Comment by bjne** [comment]
Its not really important to me to use the fromstring function, I am open to suggestions on how it should be implemented, but I also feel bytea is a bit different beast.

given a function that returns bytea, one is afaik not able to actually return binary data from that function.

Perhaps I am misunderstanding the primary use of bytea? But as you understand, I really just want the lua string to be treated as a bytea. What is your suggestion on how you would accept this implemented?

**Comment by eugwne** [comment]
int64 is one of implementation of a cutom type (lua_int64.c), it is not the best (because lua 5.3 has native support int64), but still you can get some ideas or make custom byte array type even better. And why only a string and not the array like {0x00, 0x01, 0x02 ...} ?
